### PR TITLE
the default property of a prop definition is a function, so we can do…

### PR DESCRIPTION
**Pull request by trusktr**
The default property of a prop definition can be a function, so we can do useful things like call `Math.random` or use `this` to do instance-specific things.

* [ ] Bug
* [x] Feature

## Requirements

* [x] Read the [contribution guidelines](https://github.com/skatejs/skatejs/blob/5.x/CONTRIBUTING.md).
* [ ] Wrote tests.
* [ ] Updated docs and upgrade instructions, if necessary.

## Rationale

This makes it easy to make dynamic default values. For example,

```js
const THREEColorProp = {
    attribute: { source: true, target: false }, // get the value from an attribute (but don't mirror it back)
    coerce: val => val instanceof Color ? val : new Color(val),
    deserialize: val => new Color(val), // f.e. accepts a hex "#af3eea" color string
    serialize: val => new Color( val ).getStyle(), // returns a CSS "rbg()" string

    // Let the default Color be random for each instance that has this prop
    default: () => new Color( Math.random(), Math.random(), Math.random() ), // R, G, B
}
```

## Implementation

This way the value can be a default static value, or a function that returns a value. If a function, it is called every time a default value is needed, and also called with `this` which can be useful in some cases where props are designed to work with certain classes that handle their own default values. 

Basically, it gives the end user more flexibility.

The only caveat (which we'd put in the docs) is that if the value is meant to be a function, then the function should be returned from the function:

```js
const someFunction = () => { /* ... */ }

class Foo extends withUpdate(HTMLElement) {
    static props = {
        foo: {
            // ...
            default: () => someFunction
        }
    }
}
```

## Tasks

* [ ] add to docs
* [ ] add a test

### CODE REVIEW

**Comment by trusktr** [comment]
Oops, the commit hook committed my other unstaged changes I was playing with (https://github.com/skatejs/skatejs/issues/1487).

For now just pay attention to the stuff regarding `default`, and I can remove the other stuff.

**Comment by treshugart** [comment]
If the prop default is a function, then this would be a breaking change no?

Looking down the road at v6, default props are specified as class properties or in the constructor, so you could just do this in the constructor. You can probably just do that now in v5. I know it's nice for certain props but I have a few problems with it:

- I really want prop types to do no dynamic transforms on the value.
- The only ones we really *must* do are for property serialization and deserialization; there's nothing we can do about that.
- I'm not sure how you'd type the return values and instance properties. It feels like, even if it was possible, it'd be a nasty mess of generics.

I've found, that most of the time, for default props on a pre-defined property, you're looking for a value that just "doesn't error" (tm) when being passed through the functions. I'd prefer it if types on the custom element instance were leveraged here instead and prop types were built to be resilient where values are handled internally (or typed).

**Comment by trusktr** [comment]
> If the prop default is a function, then this would be a breaking change no?

In this case no, because the normalization process converts a static value into a function that return the static value, so defining `default: 123` still works.

> Looking down the road at v6, default props are specified as class properties or in the constructor, so you could just do this in the constructor. 

Ah, interesting, I didn't realize this yet... Let me check this out...

**Comment by trusktr** [comment]
> I'd prefer it if types on the custom element instance were leveraged here instead and prop types were built to be resilient where values are handled internally (or typed).

That sounds good; props for typing, and classes defining their defaults.

**Comment by trusktr** [comment]
In my cases, the default values aren't for "not erroring", but more for nice default visuals. For example, rendering WebGL objects with default colors, default sizes, default fog value, etc, so that out of the box things look nice.

**Comment by treshugart** [comment]
This ended up making its way to `@skatejs/element` so closing.
